### PR TITLE
Falcon: batched generation

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR does three things:
1. Fixes the minimum float number added to the attention mask, in the positions the attention mask is `0`. In some numerical precisions, the numerical attention mask was getting `-inf`, which wrecked downstream computations.
2. Adds the `position_ids` input to Falcon, which is needed for proper batched generation. When it is not passed, the forward pass builds the position ids from the sequence length, which does not account for the left-padding in batched generation -- the model could still generate, but the results should be slightly better after the fix.
3. Add tests for batched generation with left padding

### CODE REVIEW

File: src/transformers/models/falcon/modeling_falcon.py [hunk 1]
```diff
@@ -99,19 +99,23 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
         self.cos_cached = self.cos_cached.type(dtype)
         self.sin_cached = self.sin_cached.type(dtype)
 
-    def cos_sin(self, seq_len: int, past_key_values_length: int, device="cpu", dtype=torch.bfloat16) -> torch.Tensor:
+    def cos_sin(self, seq_len: int, past_key_values_length: int, position_ids: torch.Tensor, device="cpu", dtype=torch.bfloat16) -> torch.Tensor:
         total_length = seq_len + past_key_values_length
         if total_length > self.seq_len_cached:
             self._set_cos_sin_cache(total_length, device, dtype)
-        return (
-            self.cos_cached[:, past_key_values_length : seq_len + past_key_values_length],
```
Comment (gante) [inline src/transformers/models/falcon/modeling_falcon.py:107]: the slicing here is equivalent to building position ids from the sequence length, without taking into account any potential left-padding

**Comment by gante** [comment]
@ArthurZucker woops, sorry, there are still tests to fix, I will ping you again when they are fixed!

**Comment by gante** [comment]
@ArthurZucker ready now

File: src/transformers/models/falcon/modeling_falcon.py [hunk 2]
```diff
@@ -415,7 +437,11 @@ def forward(
         else:
             present = None
 
-        attention_mask_float = (attention_mask * 1.0).masked_fill(attention_mask, float("-1e9")).to(query_layer.dtype)
```
Comment (gante) [inline src/transformers/models/falcon/modeling_falcon.py:418]: cc @Rocketknight1 this `1e-9` was causing problems in some numerical precisions (it would be converted to `-inf`)

Comment (Rocketknight1) [reply src/transformers/models/falcon/modeling_falcon.py:418]: Ah, my bad!

File: src/transformers/models/falcon/modeling_falcon.py [hunk 3]
```diff
@@ -415,7 +437,11 @@ def forward(
...
+        attention_mask_float = (
+            (attention_mask * 1.0)
+            .masked_fill(attention_mask, torch.finfo(query_layer.dtype).min)
+            .to(query_layer.dtype)
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/falcon/modeling_falcon.py:444]: can              fit in two lines

File: src/transformers/models/falcon/modeling_falcon.py [hunk 4]
```diff
@@ -993,8 +1037,18 @@ def prepare_inputs_for_generation(
         if past_key_values is not None:
             input_ids = input_ids[:, -1:]
 
+        # Note: versions of Falcon with alibi do not use position_ids. It is used with RoPE.
+        position_ids = kwargs.get("position_ids", None)
```
Comment (ArthurZucker) [inline src/transformers/models/falcon/modeling_falcon.py:1041]: can be added to the function args

File: src/transformers/models/falcon/modeling_falcon.py [hunk 5]
```diff
@@ -99,19 +99,40 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
         self.cos_cached = self.cos_cached.type(dtype)
         self.sin_cached = self.sin_cached.type(dtype)
 
-    def cos_sin(self, seq_len: int, past_key_values_length: int, device="cpu", dtype=torch.bfloat16) -> torch.Tensor:
+    def cos_sin(
```
Comment (ArthurZucker) [inline src/transformers/models/falcon/modeling_falcon.py:102]: I'm gonna be a bit noisy here, but this looks a LOT like the rotary embedding we have in Llama no? 
The query expansion is also supported there, not sure how much of an overhead it is to first apply rotary then expand:
```python 
        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)

        if past_key_value is not None:
            # reuse k, v, self_attention
            key_states = torch.cat([past_key_value[0], key_states], dim=2)
            value_states = torch.cat([past_key_value[1], value_states], dim=2)

        past_key_value = (key_states, value_states) if use_cache else None

        # repeat k/v heads if n_kv_heads < n_heads
        key_states = repeat_kv(key_states, self.num_key_value_groups)
        value_states = repeat_kv(value_states, self.num_key_value_groups)
```
and also storing the full size key and values is less memory efficient no? (unrelated to the PR).

**Review by ArthurZucker — COMMENTED** [review]
Thanks for adding batch support! Let's be careful with padding as we have been getting a lot of issues regarding this! If we have one reference implementation (used in Llama) would be great to re-use! 
Otherwise, LGTM

**Comment by gante** [comment]
> Let's be careful with padding as we have been getting a lot of issues regarding this!

@ArthurZucker 100% agreed! If you come across a new model, plz make sure there is a test for this 🙏

File: src/transformers/models/falcon/modeling_falcon.py (hunk 5 above)
Comment (gante) [reply src/transformers/models/falcon/modeling_falcon.py:102]: I *think* they are the same. So, we would benefit from copying the structure (at least in terms of complexity) 👍 

I would like to push it to the future, though, as I'm about to go on long holidays and I'd like to enable batched generation on Falcon :D

**Comment by gante** [comment]
@ArthurZucker suggestions applied 💪

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot! Let's add a TODO so that we don't ever forget we have to refactor this in the futur! 😉
